pkg/snowflake/tag_attachment.go

```diff
@@ -0,0 +1,107 @@
+	tagName      string
+	tagValue     string
+}
+
+type tagAttachment struct {
+	ColumnId       sql.NullString `db:"COLUMN_ID"`
+	ColumnName     sql.NullString `db:"COLUMN_NAME"`
+	Domain         sql.NullString `db:"DOMAIN"`
+	ObjectDatabase sql.NullString `db:"OBJECT_DATABASE"`
+	ObjectDeleted  sql.NullString `db:"OBJECT_DELETED"`
+	ObjectId       sql.NullString `db:"OBJECT_ID"`
+	ObjectName     sql.NullString `db:"OBJECT_NAME"`
+	ObjectSchema   sql.NullString `db:"OBJECT_SCHEMA"`
+	TagDatabase    sql.NullString `db:"TAG_DATABASE"`
+	TagId          sql.NullString `db:"TAG_ID"`
+	TagName        sql.NullString `db:"TAG_NAME"`
+	TagSchema      sql.NullString `db:"TAG_SCHEMA"`
+	TagValue       sql.NullString `db:"TAG_VALUE"`
+}
+
+// WithResourceId adds the name of the schema to the TagAttachmentBuilder
+func (tb *TagAttachmentBuilder) WithResourceId(resourceId string) *TagAttachmentBuilder {
+	tb.resourceId = resourceId
+	return tb
+}
+
+// WithObjectType adds the object type of the resource to add tag attachement to the TagAttachmentBuilder
+func (tb *TagAttachmentBuilder) WithObjectType(objectType string) *TagAttachmentBuilder {
+	tb.objectType = objectType
+	return tb
+}
+
+// WithTagValue adds the name of the tag value to the TagAttachmentBuilder
+func (tb *TagAttachmentBuilder) WithTagValue(tagValue string) *TagAttachmentBuilder {
+	tb.tagValue = tagValue
+	return tb
+}
+
+// TagAttachment returns a pointer to a Builder that abstracts the DDL operations for a tag attachment.
+//
+// Supported DDL operations are:
+//   - CREATE TAG
+//   - DROP TAG
+//
+// [Snowflake Reference](https://docs.snowflake.com/en/user-guide/object-tagging.html)
+func TagAttachment(tagId string) *TagAttachmentBuilder {
+	parsedString := strings.Replace(tagId, "\"", "", -1)
+
+	var s []string
+	if strings.Contains(parsedString, "|") {
+		s = strings.Split(strings.ToUpper(parsedString), "|")
+	} else {
+		s = strings.Split(strings.ToUpper(parsedString), ".")
+	}
+	return &TagAttachmentBuilder{
+		databaseName: s[0],
+		schemaName:   s[1],
+		tagName:      s[2],
+	}
+}
+
+// Create returns the SQL query that will set the tag on a resource.
+func (tb *TagAttachmentBuilder) Create() string {
+	//return fmt.Sprintf(`USE WAREHOUSE %v ALTER %v %v SET TAG %v = "%v"`, tb.warehouse, tb.objectType, tb.resourceId, tb.tagName, tb.tagValue)
+	return fmt.Sprintf(`ALTER %v %v SET TAG "%v"."%v"."%v" = '%v'`, tb.objectType, tb.resourceId, tb.databaseName, tb.schemaName, tb.tagName, tb.tagValue)
+}
+
+// Drop returns the SQL query that will remove a tag from a resource.
+func (tb *TagAttachmentBuilder) Drop() string {
+	return fmt.Sprintf(`ALTER %v %v UNSET TAG %v`, tb.objectType, tb.resourceId, tb.tagName)
+}
+
+func ListTagAttachments(tb *TagAttachmentBuilder, db *sql.DB) ([]tagAttachment, error) {
+	stmt := fmt.Sprintf(`SELECT * FROM SNOWFLAKE.ACCOUNT_USAGE.TAG_REFERENCES WHERE TAG_NAME = '%v' AND DOMAIN = '%v' AND TAG_VALUE = '%v'`, tb.tagName, tb.objectType, tb.tagValue)
```

Comment:
@sfc-gh-jalin  Quick question on this, from what I see in the [docs](https://docs.snowflake.com/en/sql-reference/account-usage.html) the `TAG_REFERENCES` view can have latency up to 2 hours, where as there isn't any latency using `INFORMATION SCHEMA`. I'm not super familiar with these views, but do you see a scenario where a tag has been attached but a subsequent read operation doesn't reflect the attachment because the view isn't up to date yet? This may lead to spurious diffs and some possibly unexpected behavior.

Reply:
Hi @berosen, our team did some manual testing and found that the latency is about one hour on average using the TAG_REFERENCES view. We haven't looked into the INFORMATION_SCHEMA yet but it's probably something we should look into. For now we have a retry mechanism